pkg/tbtc/heartbeat.go

```diff
@@ -69,29 +101,54 @@ func newHeartbeatAction(
 	executingWallet wallet,
 	signingExecutor heartbeatSigningExecutor,
 	proposal *HeartbeatProposal,
+	failureCounter *heartbeatFailureCounter,
+	inactivityClaimExecutor heartbeatInactivityClaimExecutor,
 	startBlock uint64,
 	expiryBlock uint64,
 	waitForBlockFn waitForBlockFn,
 ) *heartbeatAction {
 	return &heartbeatAction{
-		logger:          logger,
-		chain:           chain,
-		executingWallet: executingWallet,
-		signingExecutor: signingExecutor,
-		proposal:        proposal,
-		startBlock:      startBlock,
-		expiryBlock:     expiryBlock,
-		waitForBlockFn:  waitForBlockFn,
+		logger:                  logger,
+		chain:                   chain,
+		executingWallet:         executingWallet,
+		signingExecutor:         signingExecutor,
+		proposal:                proposal,
+		failureCounter:          failureCounter,
+		inactivityClaimExecutor: inactivityClaimExecutor,
+		startBlock:              startBlock,
+		expiryBlock:             expiryBlock,
+		waitForBlockFn:          waitForBlockFn,
 	}
 }
 
 func (ha *heartbeatAction) execute() error {
-	// TODO: When implementing the moving funds action we should make sure
-	// heartbeats are not executed by unstaking clients.
+	// Do not execute the heartbeat action if the operator is unstaking.
+	isUnstaking, err := ha.isOperatorUnstaking()
+	if err != nil {
+		return fmt.Errorf(
+			"failed to check if the operator is unstaking [%v]",
+			err,
+		)
+	}
+
+	if isUnstaking {
+		ha.logger.Warn(
+			"quitting the heartbeat action without signing because the " +
+				"operator is unstaking",
+		)
+		return nil
```

Comment:
@lukasz-zimnoch Should we consider it an error is operator is unstaking? Or leave `nil` as the returned value?

Reply:
Returning `nil` is totally fine. Unstaking operator is not an erroneous case.